opencart 3.0.0.0 - 3.0.3.1/upload/admin/controller/extension/module/smaily_for_opencart.php

```diff
@@ -85,7 +85,31 @@ public function index() {
                 'name' => $this->language->get('section_abandoned'),
             ),
         );
-
+        // Get URL for CRON links.
+        $url = new Url(HTTPS_CATALOG, $this->config->get('config_secure') ? HTTPS_CATALOG : HTTP_CATALOG);
```

Comment:
Regardless of `config_secure` value `$url->link` will always return secure store URL. It should use secure when possible and fallback to unsecure store URL.

Reply:
Sorry I don't understand this code at all. It should build new URL if HTTPS_CATALOG fails right or how do I check if secure is possible?

Reply:
Which part of the code you are struggling with?

This is just the part where you are creating an instance of a class:
```php
$url = new Url('abc/', 'xyz/');
```

Provided arguments are passed to `Url::__construct` method (see `system/library/url.php` line 25). Which assignes those values to `Url::url` and `Url::ssl` private properties:
```php
public function __construct($url, $ssl = '') {
    $this->url = $url;  // Has the value of "abc/".
    $this->ssl = $ssl;  // Has the value of "xyz/".
}
```

When you are calling
```php
$url->link('path/to/my/module', array('foo' => 'bar'), true);
```
it returns `xyz/index.php?route=path/to/my/module&foo=bar` because `Url::ssl` has a value of `xyz/` and called `link` method's 3rd argument is `true` (see `system/library/url.php` line 49 why this is happening).

If you'd call
```php
$url->link('path/to/my/module', array('foo' => 'bar'));
```
it would return `abc/index.php?route=path/to/my/module&foo=bar` because both `Url::ssl` and `link` method's 3rd argument need to be truthy in order for method to return secure URL (again see line 49 of `system/library/url.php`).

The same happens when you create an instance without assigning value to `Url::ssl` and have `true` as the 3rd argument of `link`:
```php
$url = new Url('abc/');
$url->link('path/to/my/module', array('foo' => 'bar'), true);
```
`link` would return `abc/index.php?route=path/to/my/module&foo=bar`, because even though `link`'s 3rd argument is true, `Url::ssl` does not have a truthy value which means unsecure URL is returned.

You can use class instantiation argument values and `link` method 3rd argument to control whether secure or unsecure URL should be returned.

If you have any questions regarding those examples, do not hesitate to ask. There is no such thing as stupid questions, it is all part of the learning process.

Reply:
OK. I think I got confused by the _shorthand if_ in the class instantiation and missed the HTTP**S**
I read it as $url = new Url(HTTPS_CATALOG, $this->config->get('config_secure')**)** ? HTTPS_CATALOG : HTTP_CATALOG);

Reply:
If the code gets difficult to read, then you should change it to make it more readable. Maybe extracting the second argument of class constructor could have been the solution.

With your latest changes though the readability is not an issue (I think).